Switch the `Standard` distribution for floats to [0, 1)

pitdicker opened the pull request:
See the discussion in https://github.com/rust-lang-nursery/rand/issues/416. cc @sbarral

This moves the current `Standard` distribution for generating floats to `Open01`.
`Standard` instead samples from the closed-open interval `[0, 1)` using a multiply-based method.

I did not change the Ziggurat and `Uniform` implementations, they continue to use the transmute-based methods. An implementation of `Uniform` optimized for the other method would look quite different.

I hope just those two distributions, `Standard` and `Open01`, are good enough for the 0.5 release. I am not convinced adding an explicit `ClosedOpen01` adds much benefit, but it can easily be added in the future. I am also not convinced on the usefullness of `OpenClosed01`, but added the suggestion of using `1.0 - rng.gen()` in the documentation.

dhardy commented on src/distributions/mod.rs:
```diff
@@ -275,42 +276,30 @@ impl<'a, D, R, T> Iterator for DistIter<'a, D, R, T>
+/// The `Open01` distribution provides an alternative: it generates values in
+/// the open interval (0, 1), with one less random bit. It uses the slightly
+/// faster transmute-based approach for the conversion to a floating point
+/// value.
```
"Slightly faster" isn't really correct; I just benchmarked and found no difference. You could however say *(which may be faster on some architectures)*.

dhardy commented on src/distributions/mod.rs:
````diff
@@ -275,42 +276,30 @@ impl<'a, D, R, T> Iterator for DistIter<'a, D, R, T>
-/// ```
-/// use rand::{thread_rng, Rng};
+/// `Rng::gen_range(0, 1)` also uses the faster transmute-based method, but
+/// produces values in a half-open interval just like `Standard`.
````
Just remove 'faster' here?

dhardy commented on src/distributions/mod.rs:
````diff
@@ -275,42 +276,30 @@ impl<'a, D, R, T> Iterator for DistIter<'a, D, R, T>
-/// ```
+/// As a final alternative, if you require the floating point value to never be
+/// exactly `0.0`: consider using `1.0 - rng.gen()`, which transforms the
+/// half-open `Standard` distribution to the `(0, 1]` interval.
````
I would prefer to assume any users who care know what they want:

> If you wish to sample from the `(0, 1]` half-open interval consider using `1.0 - rng.gen()`.

Looking at the code, I couldn't actually think of an alternative which samples with this precision and doesn't use an extra op (though of course one could add `ε/2` or 1 before conversion to float).

sbarral replied:
> Looking at the code, I couldn't actually think of an alternative which samples with this precision and doesn't use an extra op (though of course one could add ε/2 or 1 before conversion to float).

Explicit `OpenClosed01` does have advantages though:
* the intent is clearer and the code less cluttered,
* the `+1` integer addition before conversion is faster than the `1.0 - u` user-side workaround

dhardy replied:
I agree with you. There is also some benefit to an explicit `ClosedOpen01`, even if just for local documentation *(this value is strictly less than 1)*.

sbarral replied:
Yes, this is also something that hit me when looking at the diff: if not for @pitdicker's perceptivity, subtle bugs would have been introduced in Ziggurat and Gamma where `rng.gen()` was previously used. When I read `rng.gen()` it is always unclear to me if the author of the code did so for the convenient notation or whether the code actually relies on the guarantees regarding the `0.0` and `1.0` bounds.

pitdicker replied:
I suppose that makes for a good argument.

In this case I had it easy, as I also was the one introducing the ambiguity in https://github.com/rust-lang-nursery/rand/pull/274 :smile:.

pitdicker commented:
Changed the documentation as suggested and added back the benchmarks.

I am not against `OpenClosed01` and `ClosedOpen01`, but I don't see them as something very useful. Is it ok to merge this PR, and I leave implementing those to someone who cares more for them? This PR is tackling a breaking change, the other two are only additions.

dhardy commented:
@pitdicker can you pull [my commit](https://github.com/dhardy/rand/commits/closed_open_float_conversion)? I added `OpenClosed01` but not the other one (partly because a completely redundant distribution may cause confusion as to which to use).

I think I'm ready to merge after that.

And CI works again! :tada:

pitdicker commented:
Looks good!

vks commented:
Does it make sense to have some alias `ClosedOpen01`? This is more explicit (documenting intent) and makes code more forward compatible in case we change `Standard` again.

dhardy commented:
Yes and no: it is more explicit, but redundant and not very useful. I don't think we're likely to move `Standard` away from `[0, 1)` after all this discussion.

If you really want to see it open a PR, but I'm not sure there's sufficient motive.

vks commented:
I would prefer
```rust
let r = rng.sample(ClosedOpen01);
```
over
```rust
let r = rng.sample(Standard);  // Can never be 1.0
```
but I can't think of a use case. (It would be different if `Standard` was like `OpenClosed01`.) If that ever changes, I will open a new issue/PR.
